### PR TITLE
Support unused parameters in dynamic graph distributed

### PR DESCRIPTION
**Pull request by ForFishes**
### PR types

Bug fixes 
### PR changes

Others
### Describe

Support unused parameters in dynamic graph distributed
Currently, there are 3 situations in which dynamic graph distribution cannot be handled.
1. stop_gradient
2. detach
3. Create useless parameters in the networking statement

In these three cases, there are unused parameters, resulting in no sync operation every time allreduce is executed. Here, through the forward process, traverse the entire reverse network, find all unused var, and then mark ready in advance when calculating.

### CODE REVIEW

File: paddle/fluid/imperative/reducer.h [hunk 1]
```diff
@@ -194,6 +204,14 @@ class Reducer {
...
   std::vector<int64_t> rebuild_var_indices_;
   const std::vector<size_t> group_size_limits_;
+
+  // Following variables are to help unused vars
+  std::unordered_map<GradOpNode*, size_t> node_deps_;
+  std::unordered_map<VariableWrapper*, size_t> var_index_map_;
+  std::vector<size_t> unused_vars_;
+  bool has_marked_unused_vars_{false};
+  bool find_unused_vars_{false};
+  bool need_finalize_backward_{false};
```
Comment (hutuxian) [inline paddle/fluid/imperative/reducer.h]: `need_finalize_backward_` is little bit obscure, how about `all_group_ready`?

File: paddle/fluid/imperative/reducer.cc [hunk 2]
```diff
@@ -244,21 +246,143 @@ void Reducer::InitializeGroups(
...
+    group.all_length_ = 0;
+    group.dense_tensors_.clear();
+    group.dense_tensors_.reserve(group.pending_);
+    group.sparse_contents_ = nullptr;
   });
+
+  PADDLE_ENFORCE_EQ(
+      need_finalize_backward_, false,
+      platform::errors::PreconditionNotMet(
+          "Please note that all `forward` outputs derived from the module "
```
Comment (hutuxian) [inline paddle/fluid/imperative/reducer.cc]: What the use of `` around forward

File: paddle/fluid/imperative/reducer.cc [hunk 3]
```diff
@@ -244,21 +246,143 @@ void Reducer::InitializeGroups(
...
+  unused_vars_.clear();
+  node_deps_.clear();
+  std::queue<std::shared_ptr<GradOpNode>> q;
+  std::unordered_set<VariableWrapper *> var_visited;
+  std::unordered_set<GradOpNode *> init_nodes;
+
+  for (const auto &output : outputs) {
+    const auto &grad_node = output->GradVarBase()->GradNode();
+    if (grad_node == nullptr || output->OverridedStopGradient()) {
+      VLOG(3) << "Skip auto grad since there is no grad op for output is "
```
Comment (hutuxian) [inline paddle/fluid/imperative/reducer.cc]: for -> or?

File: paddle/fluid/imperative/reducer.h [hunk 4]
```diff
@@ -194,6 +204,14 @@ class Reducer {
   std::vector<std::shared_ptr<imperative::VarBase>> rebuild_vars_;
   std::vector<int64_t> rebuild_var_indices_;
   const std::vector<size_t> group_size_limits_;
+
+  // Following variables are to help unused vars
+  std::unordered_map<GradOpNode*, size_t> node_deps_;
+  std::unordered_map<VariableWrapper*, size_t> var_index_map_;
+  std::vector<size_t> unused_vars_;
+  bool has_marked_unused_vars_{false};
```
Comment (hutuxian) [inline paddle/fluid/imperative/reducer.h:213]: Add some comments to explain what this var control? And why we design in this way?

File: paddle/fluid/imperative/reducer.h (hunk 1 above)
Comment (ForFishes) [reply paddle/fluid/imperative/reducer.h]: ok, i will fix it.

File: paddle/fluid/imperative/reducer.cc (hunk 2 above)
Comment (ForFishes) [reply paddle/fluid/imperative/reducer.cc]: ok, i fix it.

**Review by XieYunshen — APPROVED** [review]

**Review by hutuxian — APPROVED** [review]
LGTM

**Pull request merged**